src/coreclr/nativeaot/Runtime/arm/GcProbe.S

```diff
@@ -8,6 +8,104 @@
 #include "AsmOffsets.inc"
 
 .global RhpGcPoll2
+.global RhpThrowHwEx
+
+// See PUSH_COOP_PINVOKE_FRAME, this macro is very similar, but also saves return registers
+// and accepts the register bitmask
+// Call this macro first in the method (no further prolog instructions can be added after this).
+//
+//  threadReg     : register containing the Thread* (this will be preserved).
+//  trashReg      : register that can be trashed by this macro
+//  BITMASK       : value to initialize m_dwFlags field with (register or #constant)
+.macro PUSH_PROBE_FRAME threadReg, trashReg, BITMASK
+    // Define the method prolog, allocating enough stack space for the PInvokeTransitionFrame and saving
+    // incoming register values into it.
+    PROLOG_STACK_ALLOC  4                   // Space for saved APSR
+    PROLOG_VPUSH        "{d0-d3}"           // Save floating point return registers
+    PROLOG_PUSH         "{r0-r3,lr}"        // Save volatile registers
```

Comment:
This macro is used to push a frame when a method returns. Volatile regiters are dead at that point, so no need to save. 
Except for the return registers that, although volatile, may contain live values.

Reply:
I think PUSH_PROBE_FRAME should be similar to corresponding PUSH_COOP_PINVOKE_FRAME + the same diff as between arm64 versions of PUSH_COOP_PINVOKE_FRAME and PUSH_PROBE_FRAME - i.e. PUSH_PROBE_FRAME also saves return registers and takes flags parameter.

Reply:
I wrote it as a bit of a mix between `PUSH_COOP_PINVOKE_FRAME` and the code in GcProbe.S. There are some parts which are unnecessary dead code, especially since it's only ever called with one specific BITMASK parameter.  I intend to clean it up but I wanted to fix the crashes first.

Reply:
One possibly helpful thing could be to implement arm32 version of `ForEachPossibleObjectRef` and try running smoke tests with `export DOTNET_gcConservative=1` - to see how far it can get without involving hijacking.

Some libraries tests would have problems with conservative mode, since they test/rely on deterministic finalization, but smoke tests should generally pass. Unless there are issues besides hijacking. Sometimes it is useful to be sure about that.

Reply:
I got the hijacking mostly working but I see HUGE retry counts in `SuspendAllThreads`. That happens both on QEMU and real HW. It's slightly less bad when SIGUSR1 is used instead of the real-time SIGRTMIN for the injection.

 The only failing test at the moment is `ThreadLocalStatics.TLSTesting.ThreadLocalStatics_Test` inside `DynamicGenerics`, and it is flaky and passes about 50% of the time on the Raspberry Pi.

Reply:
> HUGE retry counts

Is that because we keep hitting non-unwindable code?

Reply:
> Volatile regiters are dead at that point, so no need to save.

Fixed this part.

Reply:
> Is that because we keep hitting non-unwindable code?

I have many different theories but I don't have enough evidence to support them. I don't have reliable data about hitting non-unwindable code but I am reasonably sure that's not the case. There may be an issue with slow unwinding (again, no data to support it, but it's largely unexplored and unprofiled code paths).

On QEMU (ARM user emulation on x64 host) I see a lot of race conditions with the logic in `RhpPInvokeReturn` where `RhpWaitForGC2` is enter and it temporarily removes the `PInvokeTransitionFrame`. Whenever I stop it in the debugger in the `SuspendAllThreads` loop the return value of `CacheTransitionFrameForSuspend` was clearly `NULL` but the frame is present in the thread structure. I thought it's an artifact of how the QEMU emulation works and how it handles signals but I witnessed a very similar pattern on the Raspberry Pi.

Reply:
Here's one example of the high iteration count in `SuspendAllThreads`: https://gist.github.com/filipnavara/b42ce8459ff36d1d671e89a0b09f1144

(not sure if there's anything relevant there, just posting it for posterity)